sql/log.cc

```diff
@@ -12934,3 +13110,227 @@ void wsrep_register_binlog_handler(THD *thd, bool trx)
 }
 
 #endif /* WITH_WSREP */
+
+inline bool Binlog_commit_by_rotate::should_commit_by_rotate(
+    const MYSQL_BIN_LOG::group_commit_entry *entry) const
+{
+  binlog_cache_data *trx_cache= entry->cache_mngr->get_binlog_cache_data(true);
+  binlog_cache_data *stmt_cache=
+      entry->cache_mngr->get_binlog_cache_data(false);
+
+  if (likely(trx_cache->get_byte_position() <=
+                 opt_binlog_commit_by_rotate_threshold &&
+             stmt_cache->get_byte_position() <=
+                 opt_binlog_commit_by_rotate_threshold))
+    return false;
+
+  binlog_cache_data *cache_data= trx_cache;
+  if (unlikely(entry->using_stmt_cache && !stmt_cache->empty()))
+    cache_data= stmt_cache;
+
+  /*
+     Don't do rename if no space reserved or no nothing written in the tmp
+     file. It happens in the case binlog cache buffer is larger than
+     threshold
+   */
+  if (cache_data->file_reserved_bytes() == 0 ||
+      cache_data->cache_log.disk_writes == 0)
+    return false;
+
+  /*
+    - The binlog cache file is not encrypted in the same way with binlog, so it
+      cannot be renamed to binlog file.
+    - It is not supported to rename both statement cache and transaction cache
+      to binlog files at the same time.
+    - opt_optimize_thread_scheduling is just for testing purpose, it is usually
+      enabled, skip the disabled case to make the code simple.
+   */
+  if (encrypt_binlog || !opt_optimize_thread_scheduling ||
+      (entry->using_stmt_cache && entry->using_trx_cache &&
+       !stmt_cache->empty() && !trx_cache->empty()))
+    return false;
+
+  return true;
+}
+
+bool Binlog_commit_by_rotate::commit(MYSQL_BIN_LOG::group_commit_entry *entry)
+{
+  bool check_purge= false;
+  THD *thd= entry->thd;
+  binlog_cache_mngr *cache_mngr= entry->cache_mngr;
+  binlog_cache_data *cache_data= cache_mngr->get_binlog_cache_data(true);
+  if (unlikely(!entry->using_trx_cache || cache_data->empty()))
+    cache_data= cache_mngr->get_binlog_cache_data(false);
+
+  /* Call them before enter log_lock to avoid holding the lock long */
+  if (cache_data->sync_temp_file() || thd->wait_for_prior_commit())
+    return true;
+
+  // It will be released by trx_group_commit_leader
+  mysql_mutex_lock(&mysql_bin_log.LOCK_log);
```

Comment:
This should now read: "It will be released by trx_group_commit_with_engines()".

Reply:
fixed.